File: backend/src/main/java/com/example/graduation/entity/ChangeRequest.java
package com.example.graduation.entity;

import com.baomidou.mybatisplus.annotation.IdType;
import com.baomidou.mybatisplus.annotation.TableId;
import com.baomidou.mybatisplus.annotation.TableName;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@Data
@NoArgsConstructor
@AllArgsConstructor
@TableName("change_request")
public class ChangeRequest {

    @TableId(type = IdType.AUTO)
    private Long id;

    /** 发起变更的学生ID */
    private Long studentId;

    /** 当前生效的选题申请ID（topic_application.id） */
    private Long currentApplicationId;

    /** 变更类型：更换选题 / 更换导师 */
    private ChangeType type;

    /** 学生填写的变更原因 */
    private String reason;

    /** 学生建议的新选题（可选） */
    private Long targetTopicId;

    /** 学生建议的新导师（可选） */
    private Long targetTeacherId;

    /** 当前处理状态 */
    private ChangeStatus status;

    /** 导师审批结果（仅更换选题时使用） */
    private Decision teacherDecision;

    /** 导师审批意见 */
    private String teacherComment;

    /** 管理员审批结果（仅更换导师 / 最终确认时使用） */
    private Decision adminDecision;

    /** 管理员审批意见 */
    private String adminComment;

    private LocalDateTime createdAt;

    private LocalDateTime updatedAt;

    public enum ChangeType {
        CHANGE_TOPIC,
        CHANGE_TEACHER
    }

    public enum ChangeStatus {
        PENDING_TEACHER,
        PENDING_ADMIN,
        APPROVED,
        REJECTED,
        CANCELLED
    }

    public enum Decision {
        APPROVED,
        REJECTED
    }
}

File: frontend/src/api/changeRequest.ts
import request from './request'

export type ChangeType = 'CHANGE_TOPIC' | 'CHANGE_TEACHER'
export type ChangeStatus = 'PENDING_TEACHER' | 'PENDING_ADMIN' | 'APPROVED' | 'REJECTED' | 'CANCELLED'
export type Decision = 'APPROVED' | 'REJECTED'

export interface ChangeRequest {
  id: number
  studentId: number
  currentApplicationId: number
  type: ChangeType
  reason: string
  targetTopicId?: number
  targetTeacherId?: number
  status: ChangeStatus
  teacherDecision?: Decision
  teacherComment?: string
  adminDecision?: Decision
  adminComment?: string
  createdAt: string
  updatedAt: string
}

export interface ChangeRequestCreatePayload {
  type: ChangeType
  reason: string
  targetTopicId?: number
  targetTeacherId?: number
}

// 学生发起变更申请
export const createChangeRequest = (payload: ChangeRequestCreatePayload) => {
  return request.post<ChangeRequest>('/change-requests', payload)
}

// 学生查看自己的变更申请
export const getMyChangeRequests = () => {
  return request.get<ChangeRequest[]>('/change-requests/my')
}

// 导师查看名下待处理的“更换选题”申请
export const getTeacherPendingChangeRequests = () => {
  return request.get<ChangeRequest[]>('/change-requests/teacher/pending')
}

// 导师审批“更换选题”申请
export const submitTeacherChangeDecision = (id: number, decision: Decision, comment?: string) => {
  return request.post<void>('/change-requests/' + id + '/teacher-decision', null, {
    params: { decision, comment }
  })
}

// 管理员查看待处理的“更换导师”申请
export const getAdminPendingChangeRequests = () => {
  return request.get<ChangeRequest[]>('/change-requests/admin/pending')
}

// 管理员审批“更换导师”申请
export const submitAdminChangeDecision = (id: number, decision: Decision, comment?: string) => {
  return request.post<void>('/change-requests/' + id + '/admin-decision', null, {
    params: { decision, comment }
  })
}

